language/move-binary-format/src/unit_tests/compatibility_tests.rs

```diff
@@ -464,6 +469,145 @@ fn deprecated_add_script_visibility() {
+        check_friend_linking: true,
+        check_private_entry_linking: false,
+    }
+    .check(&updated_module, &module)
+    .is_ok());
+
+    // disallow updating signatures of private entry functions
+    assert!(Compatibility::full_check()
+        .check(&module, &updated_module)
+        .is_err());
+}
+
+#[test]
+fn public_entry_signature_change_disallowed() {
+    // Create a public entry function
+    let module = mk_module_entry(Visibility::Public as u8, true);
+    let mut updated_module = module.clone();
+    // Update the signature of the entry fun to now take a u64 argument.
+    updated_module
+        .exposed_functions
+        .get_mut(ident_str!("fn"))
+        .unwrap()
+        .parameters = vec![Type::U64];
+
+    assert!(Compatibility {
+        check_struct_and_pub_function_linking: true,
+        check_struct_layout: true,
+        check_friend_linking: true,
+        check_private_entry_linking: false
+    }
+    .check(&module, &updated_module)
+    .is_err());
+
+    assert!(Compatibility {
+        check_struct_and_pub_function_linking: true,
+        check_struct_layout: true,
+        check_friend_linking: true,
+        check_private_entry_linking: false
+    }
+    .check(&updated_module, &module)
+    .is_err());
+
+    assert!(Compatibility {
+        check_struct_and_pub_function_linking: true,
+        check_struct_layout: true,
+        check_friend_linking: true,
+        check_private_entry_linking: true
+    }
+    .check(&module, &updated_module)
+    .is_err());
+}
+
+#[test]
+fn friend_entry_signature_change_allowed() {
+    let module = mk_module_entry(Visibility::Friend as u8, true);
+    let mut updated_module = module.clone();
+    // Update the signature of the entry fun to now take a u64 argument.
+    updated_module
+        .exposed_functions
+        .get_mut(ident_str!("fn"))
+        .unwrap()
+        .parameters = vec![Type::U64];
+
+    assert!(Compatibility {
+        check_struct_and_pub_function_linking: true,
+        check_struct_layout: true,
+        check_friend_linking: false,
+        check_private_entry_linking: false
+    }
+    .check(&module, &updated_module)
+    .is_ok());
+
+    assert!(Compatibility {
+        check_struct_and_pub_function_linking: true,
+        check_struct_layout: true,
+        check_friend_linking: true,
+        check_private_entry_linking: false
+    }
+    .check(&module, &updated_module)
+    .is_err());
+
+    assert!(Compatibility {
+        check_struct_and_pub_function_linking: true,
+        check_struct_layout: true,
+        check_friend_linking: false,
+        check_private_entry_linking: true
+    }
+    .check(&module, &updated_module)
+    .is_err());
+
+    assert!(Compatibility {
+        check_struct_and_pub_function_linking: true,
+        check_struct_layout: true,
+        check_friend_linking: true,
+        check_private_entry_linking: true
+    }
+    .check(&module, &updated_module)
+    .is_err());
+}
+
```

Comment:
Let's also add tests for removing an entry function outright, and removing an `entry` modifier on a function.